certora/specs/DoubleEndedQueue.spec

```
@@ -0,0 +1,240 @@
+*/
+
+function absolute_diff(mathint a, mathint b) returns mathint {
+    return a > b ? a - b : b - a; 
+}
+
+function min_int128() returns mathint {
+    return max_uint256 / 2 + 1;
+}
+
+function max_int128() returns mathint {
+    return max_uint256 / 2;
+}
+
+// Could be broken in theory, but not in practice
+function beginLimited(int128 begin) returns bool {
+    return min_int128() < to_mathint(begin); // not inclusive so it can pushFront
+}
+
+// Could be broken in theory, but not in practice
+function endLimited(int128 end) returns bool {
+    return max_int128() > to_mathint(end); // not inclusive so it can pushBack
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: end is larger or equal than begin if values are safely assumed not to overflow.                          │
+| NOTE: This assumption make the proofs unsound, but they can't realistically overflow                                |
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant boundariesConsistency()
+    end() >= begin()
+    {
+        preserved {
+            require beginLimited(begin());
+            require endLimited(end());
+        }
+    }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: length is end minus begin                                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant lengthConsistency()
+    length() == absolute_diff(to_mathint(end()), to_mathint(begin()))
+    {
+        preserved {
+            requireInvariant boundariesConsistency;
+        }
+    }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: out of bounds is always 0                                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant outOfBoundsDefault(int128 key)
+    (
+        key < begin() &&  // Is before beginning
+        key >= end() // Is after ending
+    ) => at_(to_uint256(key)) == 0
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: empty() is length 0 and no element exists                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant emptiness(int128 key)
+    empty() <=> length() == 0 && at_(to_uint256(key)) == 0
```

be carefull with the associativity of `<=>` !

Here I would remove the `at_(to_uint256(key))` part, since its already covered by `outOfBoundsDefault`.
Instead I would check 

```suggestion
invariant emptiness()
    (empty() <=> length() == 0) && (empty() <=> begin() == end())
```

---

This comment was particularly insightful, thank you!
I applied the suggestion and I'm also still using `at_` in some places without the `@revert` but I made sure to prove `at_` only reverts when there's a query for an out of bounds value.